File: day02/main.py
def is_safe(line):
    changes = []
    for i in range(1, len(line)):
        changes.append(line[i] - line[i-1])

    all_pos = True
    all_neg = True
    conforming = True
    for c in changes:
        if (c < 0):
            all_pos = False
        if (c > 0):
            all_neg = False
        if (abs(c) < 1) or (abs(c) > 3):
            conforming = False

    return (all_pos or all_neg) and (conforming)

with open("input.txt") as fp:
    lines = [[int(x) for x in line.split(' ')] for line in fp.readlines()]
    silver = 0
    gold = 0
    for line in lines:
        pos_safe = False
        for i in range(len(line)):
            a = [line[x] for x in range(len(line)) if x != i]
            if (is_safe(a)):
                pos_safe = True
        pos_safe = pos_safe or is_safe(line)
        silver += is_safe(line)
        gold += pos_safe

    print(silver)
    print(gold)

File: day02/main.cpp
#include <algorithm>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <iterator>
#include <ranges>
#include <string>
#include <vector>
#include <ostream>

using i64 = std::int64_t;

auto tuple_difference = [](auto&& p) { const auto& [l, r] = p; return l - r; };
auto is_safe(auto&& diffs) -> bool {
    return std::ranges::fold_left(diffs, true, [](const auto& l, const auto& r) { return l && (0 < r && r < 4); })
        || std::ranges::fold_left(diffs, true, [](const auto& l, const auto& r) { return l && (-4 < r && r < 0); });
}
auto is_safe2(auto&& diffs) -> bool {
    auto res { false };
    for (const auto missing : std::views::iota(0z, std::ranges::distance(diffs))) {
        res |= is_safe(std::views::enumerate(diffs)
            | std::views::filter([missing](const auto& p) -> bool { const auto& [i, v] = p; return { i != missing }; })
            | std::views::transform([](const auto& p) { const auto& [_, v] = p; return v; })
            | std::views::adjacent<2>
            | std::views::transform(tuple_difference));
    }
    return res;
}
auto main() -> int {
    i64 answer1{0};
    i64 answer2{0};
    auto f = std::fstream{"input.txt"};
    std::string line{};
    std::vector<std::string> lines{};
    while (std::getline(f, line)) {
        auto nums = line | std::views::split(' ')
            | std::views::transform( [](const auto& s) { return std::string(&*s.begin(), std::ranges::distance(s)); })
            | std::views::transform( [](const auto& e) { return std::atoll(e.c_str()); });
        auto diffs = nums | std::views::adjacent<2> | std::views::transform(tuple_difference);
        answer1 += is_safe(diffs);
        answer2 += is_safe2(nums);
    }
    std::cout << "Answer1: " << answer1 << '\n';
    std::cout << "Answer2: " << answer2 << '\n';
    return 0;
}
